### DISCUSSION TITLE
UUID validation is unnecessarily strict / sequential UUIDs are not supported

### DISCUSSION THREAD

**Discussion by Lippur** [directus/directus#15535] [Feature Requests]
### Describe the Bug

**Updated description**

We are using Directus to manage entries in a database that is also used by a different application, which also creates records in the tables.
The UUIDs are generated using the [https://github.com/ulid/spec](ULID) algorithm and stored as UUIDs in Postgres. The result is a valid UUID for Postgres and pretty much every other application. However, the generated UUIDs do not necessarily have the expected version characters in the third and fourth blocks, therefore they are not strictly RFC4122 compliant. Therefore, the validation in Directus fails for these UUIDs, and results in a "Permission denied" message for records with an "invalid" UUID. 

However, there are many valid reason for generating UUIDs which do not strictly match the spec. Fully random UUIDs create fragmentation in the database and make indexing less efficient. It is not uncommon to generate sequential UUIDs with algorithms such as the one described above.

As Directus does not rely on the version information from the UUIDs, I suggest that the validation be relaxed to accommodate these use cases, and prevent hard-to-debug errors. 

---
Old description:

We're running into a weird bug where Directus shows "Page not found" with "Access forbidden" in the network response when accessing certain entires, but works fine with others. There is seemingly no difference between the different records. For example, we could have 10 entries in a collection, and 3 would work while the rest don't. One thing we noticed is that entries created through Directus always seem to work. Only externally created entries have this issue, but not all of them. This happens while logged in with any user, no matter the user's permissions, even with an administrator account. The issue always happens on the same records. If a given record doesn't work, it always doesn't work.

This happens across 3 different environments, so it's not an isolated occurrence. The user that Directus uses to connect to the database is the database owner with full permissions.  

This is an example for 3 rows from the database:
<img width="1440" alt="image" src="https://user-images.githubusercontent.com/22624090/189363197-6cebf706-8233-424a-a76a-39068172e713.png">

`Test vendor` and `Test vendor 2` work fine, `Test vendor 1` throws the error.  

`directus_collections`:  
|collection|icon|note|display_template|hidden|singleton|translations|archive_field|archive_app_filter|archive_value|unarchive_value|sort_field|accountability|color|item_duplication_fields|sort|group|collapse|
|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
|Vendor||||false|false|||false||||||[]|||open|

`directus_fields`:
|id|collection|field|special|interface|options|display|display_options|readonly|hidden|sort|width|translations|note|conditions|required|group|validation|validation_message|
|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
|4|Vendor|Name||input|"{""clear"":true}"|raw||false|false||full||||true||||
|5|Vendor|Created|date-created|datetime||datetime|"{""format"":""short""}"|true|false||full||||false||||
|6|Vendor|Updated|date-updated,date-created|datetime|"{""includeSeconds"":true}"|datetime|"{""relative"":true}"|true|false||full||||false||||
|7|Vendor|Revision|uuid|||||true|true||full||||false||||
|3|Vendor|Id|uuid|input||raw||true|false||full||||false||||
|15|Vendor|TestField||input||||false|false||full||||false||||

Listing works fine:
<img width="1157" alt="image" src="https://user-images.githubusercontent.com/22624090/189368028-e1d9c530-2b5f-4360-98f3-55a5b8b3a471.png">

I'd be happy to provide extra information or try out things to get this solved.

### To Reproduce

1. Create a row in a collection with a UUID primary key which does not validate according to RFC4122, but which is still a valid UUID from the database perspective, e.g. 01826e00-c8f3-acf8-fb5c-f86bae5ce9f2
2. Go to the edit page for the entry in Directus
3. "Page not found" is displayed

### Errors Shown

Directus logs with `log_level` set to trace:
```
12:57:14 🔍 [11.973ms] select "ip_access" from "directus_roles" where "id" = $1 limit $2 [8d7e8b2e-85b3-4414-8772-b26baba42505, 1]
12:57:14 🔍 [8.991ms] select "ip_access" from "directus_roles" where "id" = $1 limit $2 [8d7e8b2e-85b3-4414-8772-b26baba42505, 1]
12:57:14 🔍 [3.848ms] select "ip_access" from "directus_roles" where "id" = $1 limit $2 [8d7e8b2e-85b3-4414-8772-b26baba42505, 1]
12:57:14 🐛 You don't have permission to access this.
12:57:14 ✨ request completed GET 403 /items/Vendor/01826e00-c8f3-acf8-fb5c-f86bae5ce9f2 63ms
```

UI:
<img width="1008" alt="image" src="https://user-images.githubusercontent.com/22624090/189361694-7e72aa31-6b83-4440-b420-00575d4af60a.png">

API response:
<img width="754" alt="image" src="https://user-images.githubusercontent.com/22624090/189361977-386b940d-0b4b-436e-a992-7707fee08a39.png">
```
{"errors":[{"message":"You don't have permission to access this.","extensions":{"code":"FORBIDDEN"}}]}
```

### What version of Directus are you using?

9.17.4

### What version of Node.js are you using?

16.17.0

### What database are you using?

Postgres 14

### What browser are you using?

Chrome, Edge, Firefox

### How are you deploying Directus?

Docker, same issue happens locally

**Comment by Lippur** [comment]
I just changed the primary key of a row that wasn't working before to a new random UUID through the database, and it started working, no more access forbidden error for that specific row.  
Obviously, we cannot go around changing the IDs of records when they don't work, but maybe this helps troubleshoot the issue.

**Comment by br41nslug** [comment]
The "Page not found"/"Permission denied" error occurs when the routing fails to match the parameter to a valid primaryKey. 

> Create a record in the database externally, without going through Directus

How are you specifically inserting these UUIDs? to have the correct data format you'll need to cast it to UUID when inserting `'292a485f-a56a-4938-8f1a-bbbbbbbbbbb1'::UUID`

**Comment by Lippur** [comment]
> The "Page not found"/"Permission denied" error occurs when the routing fails to match the parameter to a valid primaryKey.
> 
> > Create a record in the database externally, without going through Directus
> 
> How are you specifically inserting these UUIDs? to have the correct data format you'll need to cast it to UUID when inserting `'292a485f-a56a-4938-8f1a-bbbbbbbbbbb1'::UUID`

They are inserted by a .NET application using EF Core. Every entry is inserted in the same way, so I don't see how it would make sense for some to be correct and some incorrect.

To try it out, I just tried inserting an entry manually with the explicit UUID casting:
```
INSERT INTO public."Vendor" ("Id", "Name", "Created", "Updated", "Revision")
VALUES ('01832191-e9d8-d254-4c71-b62986a5c844'::UUID, 'Another one', '2022-09-12 13:41:33.952000 +00:00',
        '2022-09-12 13:41:35.119000 +00:00', '01832191-ea12-2790-669c-1ffe8134ad65'::UUID);
```

This still gives a "Permission denied" in Directus.

```
GET http://localhost:8055/items/Vendor/01832191-e9d8-d254-4c71-b62986a5c844

{"errors":[{"message":"You don't have permission to access this.","extensions":{"code":"FORBIDDEN"}}]}
```

**Comment by Lippur** [comment]
On further testing, it does appear that this is related to the UUID format. Directus rejects UUIDs that _are_ valid UUIDs in Postgres (and most other applications), but which do not match the regex: `[0-9a-f]{8}-[0-9a-f]{4}-[0-5][0-9a-f]{3}-[089ab][0-9a-f]{3}-[0-9a-f]{12}`. In particular, the version information in the third and fourth block was "incorrect" in the UUIDs that failed.  I could "cure" a broken UUID by changing the first character of the third block to 4 and the first character of the fourth block to 8.

The application that generates these UUIDs uses the [ULID](https://github.com/ulid/spec) algorithm and stores it as a UUID.  

I see no reason why Directus should reject these UUIDs - the version information in the UUID is not really relevant for Directus, and this simply breaks things in a very unintuitive way.

**Comment by br41nslug** [comment]
The uuid used in your example does not validate according to the library we're using that might explain why it doesnt always happen
![image](https://user-images.githubusercontent.com/9389634/189647698-61c02ddb-5692-421c-b99c-e054a772c77b.png)

Can you make sure you're generating valid v4 UUIDs

> The application that generates these UUIDs uses the [ULID](https://github.com/ulid/spec) algorithm and stores it as a UUID.

Directus officially only supports UUIDv4 but v1 and v5 should work aswell. ULID however does not always generate valid uuid's following these standards and thus is not supported

**Comment by Lippur** [comment]
> The uuid used in your example does not validate according to the library we're using that might explain why it doesnt always happen ![image](https://user-images.githubusercontent.com/9389634/189647698-61c02ddb-5692-421c-b99c-e054a772c77b.png)
> 
> Can you make sure you're generating valid v4 UUIDs

See my comment above. Is there any reason why the UUID needs to be strictly RFC4122 compliant?
We use sequential UUIDs because it greatly helps with indexing and prevents fragmentation. It is not uncommon to use algorithms that generate sequential UUIDs (ULID and NewId are examples), and they _are_ valid values in the database. Since Directus (probably) does not depend on this version information, couldn't the validation be relaxed to support these use cases?  

I updated the issue description to reflect this

**Comment by br41nslug** [comment]
> See my comment above. Is there any reason why the UUID needs to be strictly RFC4122 compliant?

Im not sure but thats currently what it is. I'm moving this to a feature request to support more UUID formats 😄

**Comment by robertbakker1** [comment]
I've also run into this issue. We are using ULIDs which do not pass the validation in the package 'uuid-validate' (https://github.com/microsoft/uuid-validate). The package is now also archived by Microsoft. 
Please reconsider changing to a more relaxed one (which is also maintained) or roll an own validate function (I see no reason for it to be very strict).

**Comment by peauc** [comment]
We have encountered this issue on the latest version of Directus, with the help of [br41nslug](https://github.com/br41nslug) I implemented a workaround as described [in this post](https://github.com/directus/directus/issues/16938). 

I have free time this week to tackle this issue, can I prepare an implementation or at least a plan to fix this issue?

**Comment by peauc** [comment]
@br41nslug, would it be acceptable if I drafted a PR for this issue? We cannot change the uuid of our table and cannot access them.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Reply by Lippur** [reply]
The "R&D" required is to replace a deprecated plugin to fix a bug which not only completely prevents integration with any application using modern UUIDs, but also results in an error that gives absolutely no information about the actual issue...
It's no wonder more people haven't upvoted it, they probably never managed to figure out what's causing the issue, since the error is so uninformative and misleading. And wouldn't expect to find an obvious bugfix categorised as a feature request.

**Reply by rijkvanzanten** [reply]
> completely prevents integration with any application using modern UUIDs

Odd statement! In your original post, you clearly point out that you're using non-spec compliant UUIDs: "_the generated UUIDs do not necessarily have the expected version characters in the third and fourth blocks, therefore they are not strictly RFC4122 compliant. Therefore, the validation in Directus fails for these UUIDs"_. The current validation checks for UUID validity. If the UUID is invalid, it doesn't validate. I'd say it's a bit of a stretch to then generalize those non-standard UUIDs as "modern" without any other citations 

> It's no wonder more people haven't upvoted it, they probably never managed to figure out what's causing the issue, since the error is so uninformative and misleading.

It's really not that black and white. There's two way more likely reasons: 
1) People generally don't have the same specific use case as yourself, and therefore don't care, or (and this one is more likely IMO):
2) This feature request didn't show up on GitHub for a lot of users, as GitHub only shows the newest 1000 discussions. Very possible that people most likely didn't even realize it existed or found this feature request through search before.

The goal of this cleanup of older low-popularity feature requests is to get the list of open feature requests down to a point where people can actually see/explore them all, and easily find the one that talks about the problem they're facing. That way, we prevent a ton of duplicates and feature requests that go stale, even though they're valuable 👍🏻 

Groetjes!

**Reply by Lippur** [reply]
I appreciate that the current check is for UUID compliance with the RFC4122. But Directus does not actually need the UUIDs to be compliant with the RFC - the UUIDs are perfectly valid both as values in UUID columns in the database and for use as identifiers in the code. So the validity check right now is effectively returning a false negative result. There might be reasons to limit the UUIDs to comply with the spec, such as needing to extract the version information, but these reasons do not apply to Directus as far as I can tell.  

It's hard to see this issue marked as a feature request and closed for not being popular enough, since it is effectively a complete deal-breaker for using Directus in our projects, just as it is for the multiple other users in this discussion who ran into this and managed to narrow the issue down to being caused by these UUIDs. This use case is much more common than you might think, especially in the .NET world, where the two most popular packages providing these kinds of UUIDs, [NewId](https://www.nuget.org/packages/NewId) and [Ulid](https://www.nuget.org/packages/Ulid), have nearly 40 million downloads combined. That's not a highly specific use case, it's best practice. These millions of users are currently excluded from integrating Directus altogether, since they run into an error almost immediately, and would probably never bother to track down an issue like this, especially with the error message being clearly wrong (access forbidden).
 
I understand that you don't want to waste time on feature requests that are not popular, but I implore you to at the very least add a more descriptive error message so people know what's going wrong, or just remove the silly validation which serves absolutely no purpose but to prevent the use of more efficient UUIDs. It's a tiny change that makes a huge difference for those it affects.

**Reply by peauc** [reply]
I'm a bit surprised to see this issue closed.

As @Lippur said, this UUID issue is starting to create problems on our side and i'm sure a lot of people encountered this issue and relegated it to Directus being buggy and moved on.
We changed a few columns in DB from UUID to varchar(36) to make Directus play nice with some of our older tables but we won't be able to do it for some of our newer tables.

I guess we'll be looking to make a PR or at least a fork.
